include/shark/Algorithms/Trainers/LogisticRegression.h

```diff
@@ -0,0 +1,279 @@
+			if(m_bias){
+				RealVector mask(dim,1.0);
+				subrange(mask,dim - numOutputs, dim).clear();
+				regularizer.setMask(mask);
+			}
+			error.setRegularizer(m_lambda2, &regularizer);
+		}
+		
+		//no l1-regularization needed -> simple case
+		if(m_lambda1 == 0){
+			LBFGS optimizer;
+			error.init();
+			optimizer.init(error);
+			RealVector lastPoint = optimizer.solution().point;
+			while(norm_inf(optimizer.derivative()) > m_accuracy){
+				optimizer.step(error);
+				//if no progress has been made, something is wrong or we have numerical problems
+				//=> abort.
+				if(norm_sqr(optimizer.solution().point - lastPoint) == 0) break;
+				noalias(lastPoint) = optimizer.solution().point;
+			}
+			model.setParameterVector(lastPoint);
+			return;
+		}
+		
+		//l1-regularization is more painful.
+		//we transform the l1-regularization |w|
+		// by adding two sets of parameters, w=u-v , u >= 0, v>=0 and |w| = 1^Tu +1^Tv
+		// the resulting function is differentiable, however we have added constraints
+		L1Reformulation function(&error, m_lambda1, dim - m_bias * numOutputs);
+		LBFGS optimizer;
+		function.init();
+		optimizer.init(function);
+		RealVector lastPoint = optimizer.solution().point;
+		for(;;){
+			//check whether we are done
+			bool optimal= true;
+			auto const& derivative = optimizer.derivative();
+			for(std::size_t i = 0; i != lastPoint.size(); ++i){
+				if(lastPoint(i) < 1.e-13 && -derivative(i) > m_accuracy){//coordinate on constraint and derivative pushes away from constraint
+					optimal = false;
+					break;
+				}else if(lastPoint(i) > 1.e-13 && std::abs(derivative(i)) > m_accuracy){//free coordinate and derivative is not close to 0
+					optimal = false;
+					break;
+				}
+			}
+			if(optimal)
+				break;
+			
+			
+			
+			optimizer.step(function);
+			//if no progress has been made, something is wrong or we have numerical problems
+			//=> abort.
+			if(norm_sqr(optimizer.solution().point - lastPoint) == 0) break;
+			noalias(lastPoint) = optimizer.solution().point;
+		}
+		
+		std::size_t n = dim - m_bias * numOutputs;
+		//construct parameter vector from solution
+		RealVector param = (subrange(lastPoint,0,n) - subrange(lastPoint,n, 2 * n)) | subrange(lastPoint,2*n,lastPoint.size());
+		model.setParameterVector(param);
+	}
+
+private:
+	bool m_bias; ///< whether to train with the bias parameter or not
+	double m_lambda1;             ///< l1-regularization parameter
+	double m_lambda2;             ///< l2-regularization parameter
+	double m_accuracy;           ///< gradient accuracy
+
+	class L1Reformulation: public SingleObjectiveFunction{
+	public:
+		L1Reformulation(ErrorFunction* error, double lambda1, std::size_t regularizedParams)
+		: mep_error(error), m_lambda1(lambda1), m_regularizedParams(regularizedParams){
+			m_features |= CAN_PROPOSE_STARTING_POINT;
+			m_features |= HAS_FIRST_DERIVATIVE;
+			
+			std::size_t dim = numberOfVariables();
+			double unconstrained = std::numeric_limits<double>::max();
```

Comment:
In order to avoid infinite values, I prefer 1e100 for this purpose, since we can do calculations with it (i.e., adding epsilon) without having to deal with infinite values. Actual infinity is an alternative, but that may not be portable to non-IEEE complient systems.